src/main/java/org/opensearch/ad/transport/PMMLResultTransportAction.java

```diff
@@ -0,0 +1,137 @@
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package org.opensearch.ad.transport;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.support.ActionFilters;
+import org.opensearch.action.support.HandledTransportAction;
+import org.opensearch.ad.breaker.ADCircuitBreakerService;
+import org.opensearch.ad.common.exception.AnomalyDetectionException;
+import org.opensearch.ad.common.exception.LimitExceededException;
+import org.opensearch.ad.constant.CommonErrorMessages;
+import org.opensearch.client.node.NodeClient;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.ml.client.MachineLearningClient;
+import org.opensearch.ml.client.MachineLearningNodeClient;
+import org.opensearch.ml.common.dataframe.ColumnMeta;
+import org.opensearch.ml.common.dataframe.DataFrame;
+import org.opensearch.ml.common.dataframe.DataFrameBuilder;
+import org.opensearch.ml.common.dataframe.Row;
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.TransportService;
+
+public class PMMLResultTransportAction extends HandledTransportAction<PMMLResultRequest, PMMLResultResponse> {
+    private static final Logger LOG = LogManager.getLogger(PMMLResultTransportAction.class);
+    private ADCircuitBreakerService adCircuitBreakerService;
+    // AD's client is a node client, but we only cast it here to be used with the ML plugin
+    private final NodeClient client;
+
+    // Right now we (only) require/take "outlier" and "decisionFunction" fields out of the prediction result
+    private static String OUTLIER_FIELD = "outlier";
+    private static String DECISION_FUNCTION_FIELD = "decisionFunction";
+
+    @Inject
+    public PMMLResultTransportAction(
+        ActionFilters actionFilters,
+        TransportService transportService,
+        ADCircuitBreakerService adCircuitBreakerService,
+        NodeClient client
+    ) {
+        super(PMMLResultAction.NAME, transportService, actionFilters, PMMLResultRequest::new);
+        this.adCircuitBreakerService = adCircuitBreakerService;
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(Task task, PMMLResultRequest request, ActionListener<PMMLResultResponse> listener) {
+
+        if (adCircuitBreakerService.isOpen()) {
+            listener.onFailure(new LimitExceededException(request.getAdID(), CommonErrorMessages.MEMORY_CIRCUIT_BROKEN_ERR_MSG));
+            return;
+        }
+
+        try {
+            MachineLearningClient mlClient = new MachineLearningNodeClient(client);
```

Comment:
Can we use singleton of `MachineLearningClient` to avoid create new client for every request? You can refer to `AnomalyDetectorPlugin#createComponents`

Reply:
I have added this in this PR's "Description - Future tasks and improvements - 1"

Reply:
Sure, please add TODO in code as well, so others know we should tune this part when read code.